File: 2023/src/day24.zig
const std = @import("std");
const ascii = @import("std").ascii;
const utils = @import("./utils.zig");

const fileData = @embedFile("./files/day24.txt");
const fileDataEx = @embedFile("./files/day24ex.txt");

const Regex = @import("zig-regex").Regex;

const Allocator = std.mem.Allocator;

fn solve2x2(a: f64, b: f64, c: f64, d: f64, e: f64, f: f64) ?[2]f64 {
    //std.debug.print("Solving:\n {} {} | t0 | {} \n {} {} | t1 | {} \n", .{ a, b, e, c, d, f });
    var det = (a * d) - (b * c);
    if (det == 0) {
        return null;
    }
    var ans = [2]f64{ 0, 0 };
    ans[0] = (d * e - b * f) / det;
    ans[1] = (-c * e + a * f) / det;
    return ans;
}

const Stone = struct {
    name: []const u8,
    pos: [3]i64,
    speed: [3]i64,

    pub fn init(name: []const u8) Stone {
        return Stone{ .name = name, .pos = [3]i64{ 0, 0, 0 }, .speed = [3]i64{ 0, 0, 0 } };
    }

    pub fn intersect(self: Stone, other: Stone, area: [2]f64) bool {
        var t_f = solve2x2(@floatFromInt(self.speed[0]), @floatFromInt(-other.speed[0]), @floatFromInt(self.speed[1]), @floatFromInt(-other.speed[1]), @floatFromInt(other.pos[0] - self.pos[0]), @floatFromInt(other.pos[1] - self.pos[1]));
        //std.debug.print("Stone {any} and {any} intersect on {any}\n", .{ self.speed, other.speed, t_f });
        if (t_f == null) {
            return false;
        }
        var t = t_f.?;
        if (t[0] >= 0 and t[1] >= 0) {
            var insideArea = true;
            var pos_intersect: [2]f64 = undefined;
            for (0..2) |ix| {
                pos_intersect[ix] = t[0] * @as(f64, @floatFromInt(self.speed[ix])) + @as(f64, @floatFromInt(self.pos[ix]));
                insideArea = insideArea and (pos_intersect[ix] >= area[0] and pos_intersect[ix] <= area[1]);
            }
            return insideArea;
        }
        return false;
    }
};

fn printStringHashMap(data: std.StringHashMap(bool)) void {
    var it = data.iterator();
    while (it.next()) |item| {
        std.debug.print("{s},", .{item.key_ptr.*});
    }
}

fn parse_stone(
    line: []const u8,
    counter: usize,
    allocator: Allocator,
    stones: *std.ArrayList(Stone),
) !void {
    var re = try Regex.compile(allocator, "(\\d+), (\\d+), (\\d+) @ ([0-9-]+), ([0-9-]+), ([0-9-]+)");
    var captures = try re.captures(line);
    if (captures != null) {
        var stone = Stone.init(try std.fmt.allocPrint(allocator, "{}", .{counter}));
        for (0..3) |ix| {
            stone.pos[ix] = try std.fmt.parseInt(i64, captures.?.sliceAt(ix + 1).?, 10);
            stone.speed[ix] = try std.fmt.parseInt(i64, captures.?.sliceAt(ix + 4).?, 10);
        }
        try stones.append(stone);
    } else {
        std.debug.print("{s} not found\n", .{line});
        unreachable;
    }
}

fn solve(data: anytype, example: bool) !void {
    var arena_state = std.heap.ArenaAllocator.init(std.heap.c_allocator);
    defer arena_state.deinit();
    const allocator = arena_state.allocator();

    var splitsLines = std.mem.split(u8, data, "\n");

    var stones = std.ArrayList(Stone).init(allocator);
    var counter: usize = 0;
    while (splitsLines.next()) |l| : (counter += 1) {
        try parse_stone(l, counter, allocator, &stones);
    }

    std.debug.print("{} stones found \n", .{stones.items.len});

    var search_area = if (example) [2]f64{ 7, 27 } else [2]f64{ 200000000000000, 400000000000000 };
    var part1: i32 = 0;
    for (stones.items, 1..) |stone, ix| {
        for (stones.items[ix..]) |other| {
            if (stone.intersect(other, search_area)) {
                part1 += 1;
            }
        }
    }
    std.debug.print("Part 1 = {} \n", .{part1});
}

pub fn main() !void {
    const argv = std.os.argv;
    var example: bool = false;
    std.log.info("Hello day 24: {s}", .{argv});
    for (argv) |args| {
        if (std.mem.eql(u8, args[0..2], "ex")) {
            example = true;
        }
    }

    try solve(if (example) fileDataEx else fileData, example);
}

File: 2023/src/day24.py
import argparse
import copy
import math
import numpy as np
import time
import logging
import sympy


def intersect2D(p1, p2, r_lim):
    x1, v1 = p1
    x2, v2 = p2
    a = np.array([[v1[0], -v2[0]], [v1[1], -v2[1]]])
    b = np.array([x2[0]-x1[0], x2[1]-x1[1]])
    try:
        t = np.linalg.solve(a, b)
    except:
        logging.debug(f"{p1} and {p2}: dont intersect")
        return False

    intersect = t[0] * v1 + x1
    if t[0] < 0 or t[1] < 0:
        logging.debug(f"{p1} and {p2}: intersect in the past")
        return False

    if intersect[0] >= r_lim[0] and intersect[0] <= r_lim[1] and intersect[1] >= r_lim[0] and intersect[1] <= r_lim[1]:
        logging.debug(f"{p1} and {p2}: intersect inside area ({intersect})")
        return True

    logging.debug(f"{p1} and {p2}: intersect outside area ({intersect})")
    return False


def main():
    parser = argparse.ArgumentParser(description="Day24")
    parser.add_argument("--ex", help="Run on example", action='store_true')
    parser.add_argument("--log", dest="logLevel", choices=['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'],
                        help="Set the logging level", default="WARNING")
    args = parser.parse_args()
    logging.basicConfig(level=args.logLevel,
                        format='%(levelname)s: %(message)s')

    file = "./src/files/day24.txt"
    search_area = [200000000000000, 400000000000000]
    if args.ex:
        file = "./src/files/day24ex.txt"
        search_area = [7, 27]

    stones = []
    with open(file, "r") as f:
        for il, line in enumerate(f.readlines()):
            a, b = line.strip().split(" @ ")
            pos = np.array([int(x) for x in a.split(", ")])
            speed = np.array([int(x) for x in b.split(", ")])
            stones.append((pos, speed))

    part1 = 0
    for il, p1 in enumerate(stones):
        for p2 in stones[il+1:]:
            if intersect2D(p1, p2, search_area):
                part1 += 1
    print(f"Part 1 = {part1}")

    x, y, z, vx, vy, vz, t1, t2, t3 = sympy.symbols('x y z vx vy vz t1 t2 t3')
    equations = []
    for (pi, vi), t in zip(stones[:3], [t1, t2, t3]):
        for i, p, v in zip([0, 1, 2], [x, y, z], [vx, vy, vz]):
            equations.append(p-pi[i]-t*vi[i] - t*v)

    ans = sympy.solve(equations)[0]
    logging.debug(ans)
    part2 = ans[x]+ans[y]+ans[z]
    print(f"Part 2 = {part2}")


if __name__ == "__main__":
    main()
